src/Swarm/Game/Step.hs

```diff
@@ -1240,10 +1240,47 @@ execConst c vs s k = do
         inv <- use robotInventory
         return $ Out (VInt (fromIntegral $ countByName name inv)) s k
       _ -> badConst
+    Scout -> case vs of
+      [VDir d] -> do
+        rMap <- use robotMap
+        myLoc <- use robotLocation
+        heading <- deriveHeading d
+        botsByLocs <- use robotsByLocation
+        selfRid <- use robotID
+
+        -- Includes the base location, so we exclude the base robot later.
+        let locsInDirection :: [Location]
+            locsInDirection = take maxScoutRange $ iterate (.+^ heading) myLoc
+
+        let hasOpaqueEntity =
+              fmap (maybe False (`hasProperty` E.Opaque)) . entityAt
+
+        let hasVisibleBot :: Location -> Bool
+            hasVisibleBot = any botIsVisible . IS.toList . excludeSelf . botsHere
+             where
+              excludeSelf = (`IS.difference` IS.singleton selfRid)
+              botsHere loc = M.findWithDefault mempty loc botsByLocs
+              botIsVisible = maybe False canSee . (`IM.lookup` rMap)
+              canSee = not . (^. robotDisplay . invisible)
+
+        -- A robot on the same cell as an opaque entity is considered hidden.
+        let evalLocation :: Bool -> Location -> Maybe Bool
```

Comment:
What does the `Maybe Bool` represent?  The name `evalLocation` does not really give any hints.  I wonder if it would make the code more readable to use a custom data type like `data ScoutStatus = Opaque | VisibleBot | Clear`.

Reply:
Added some code comments.